ta/pkcs11/src/object.c

```diff
@@ -409,3 +409,591 @@ enum pkcs11_rc entry_destroy_object(struct pkcs11_client *client,
+		/*
+		 * We assume that if size is 0, pValue was NULL, so we return
+		 * the size of the required buffer for it (3., 4.)
+		 */
+		rc = get_attribute(obj->attributes, cli_ref->id,
+				   cli_ref->size ? cli_ref->data : NULL,
+				   &(cli_ref->size));
+		/* Check 2. */
+		switch (rc) {
+		case PKCS11_CKR_OK:
+			break;
+		case PKCS11_RV_NOT_FOUND:
+			cli_ref->size = PKCS11_CK_UNAVAILABLE_INFORMATION;
+			attr_type_invalid = 1;
+			break;
+		case PKCS11_CKR_BUFFER_TOO_SMALL:
+			buffer_too_small = 1;
+			break;
+		default:
+			rc = PKCS11_CKR_GENERAL_ERROR;
+			goto out;
+		}
+	}
+
+	/*
+	 * If case 1 applies to any of the requested attributes, then the call
+	 * should return the value CKR_ATTRIBUTE_SENSITIVE. If case 2 applies to
+	 * any of the requested attributes, then the call should return the
+	 * value CKR_ATTRIBUTE_TYPE_INVALID. If case 5 applies to any of the
+	 * requested attributes, then the call should return the value
+	 * CKR_BUFFER_TOO_SMALL. As usual, if more than one of these error codes
+	 * is applicable, Cryptoki may return any of them. Only if none of them
+	 * applies to any of the requested attributes will CKR_OK be returned.
+	 */
+
+	rc = PKCS11_CKR_OK;
+	if (attr_sensitive)
+		rc = PKCS11_CKR_ATTRIBUTE_SENSITIVE;
+	if (attr_type_invalid)
+		rc = PKCS11_CKR_ATTRIBUTE_TYPE_INVALID;
+	if (buffer_too_small)
+		rc = PKCS11_CKR_BUFFER_TOO_SMALL;
+
+	/* Move updated template to out buffer */
+	TEE_MemMove(out->memref.buffer, template, out->memref.size);
+
+	DMSG("PKCS11 session %"PRIu32": get attributes %#"PRIx32,
+	     session->handle, object_handle);
+
+out:
+	TEE_Free(template);
+	template = NULL;
+
+	return rc;
+}
+
+uint32_t entry_get_object_size(struct pkcs11_client *client,
+			       uint32_t ptypes, TEE_Param *params)
+{
+        const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_MEMREF_OUTPUT,
+						TEE_PARAM_TYPE_NONE);
+	TEE_Param *ctrl = params;
+	TEE_Param *out = params + 2;
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	struct serialargs ctrlargs = { };
+	struct pkcs11_session *session = NULL;
+	uint32_t object_handle = 0;
+	struct pkcs11_object *obj = NULL;
+	uint32_t obj_size = 0;
+
+	if (!client || ptypes != exp_pt)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	serialargs_init(&ctrlargs, ctrl->memref.buffer, ctrl->memref.size);
+
+	rc = serialargs_get_session_from_handle(&ctrlargs, client, &session);
+	if (rc)
+		return rc;
+
+	rc = serialargs_get(&ctrlargs, &object_handle, sizeof(uint32_t));
+	if (rc)
+		return rc;
+
+	if (serialargs_remaining_bytes(&ctrlargs))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	obj = pkcs11_handle2object(object_handle, session);
+	if (!obj)
+		return PKCS11_CKR_OBJECT_HANDLE_INVALID;
+
+	rc = check_access_attrs_against_token(session, obj->attributes);
+	if (rc)
+		return PKCS11_CKR_OBJECT_HANDLE_INVALID;
+
+	if (out->memref.size != sizeof(uint32_t))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	assert(obj->attributes);
```

Comment:
I don't think this has any value in here. We have already accessed `obj->attributes` before this.

Reply:
ok, we can remove.

Reply:
Removed